src/lsm/grid.zig

```diff
@@ -768,33 +707,102 @@ pub fn GridType(comptime Storage: type) type {
+            read: *Grid.ReadRepair,
+            block: BlockPtr,
+            address: u64,
+            checksum: u128,
+        ) void {
+            assert(address > 0);
+
+            assert(grid.superblock.opened);
+            // We try to read the block even when it is free — if we recently released it,
+            // it might be found on disk anyway.
+            maybe(grid.superblock.free_set.is_free(address));
+            // The caller will not attempt to help another replica repair a block that
+            // we are already trying to repair ourselves.
+            assert(!grid.faulty(address, null));
+            maybe(grid.writing(address, null) == .init);
+
+            read.* = .{
+                .callback = callback,
+                .address = address,
+                .checksum = checksum,
+                .block = block,
+                .grid = grid,
+            };
+
+            grid.on_next_tick(read_block_repair_tick_callback, &read.next_tick);
+        }
+
+        fn read_block_repair_tick_callback(next_tick: *Grid.NextTick) void {
+            const read = @fieldParentPtr(ReadRepair, "next_tick", next_tick);
+            const grid = read.grid;
+
+            if (grid.cache.get_index(read.address)) |cache_index| {
+                const cache_block = grid.cache_blocks[cache_index];
+
+                const header = mem.bytesAsValue(vsr.Header, cache_block[0..@sizeOf(vsr.Header)]);
+                assert(header.op == read.address);
+                if (constants.verify) grid.verify_read(read.address, cache_block);
+
+                if (header.checksum == read.checksum) {
+                    stdx.copy_disjoint(.inexact, u8, read.block, cache_block[0..header.size]);
+                    read.callback(read, {});
+                } else {
+                    read.callback(read, error.BlockNotFound);
+                }
+            } else {
+                grid.superblock.storage.read_sectors(
+                    read_block_repair_callback,
+                    &read.completion,
+                    read.block,
+                    .grid,
+                    block_offset(read.address),
+                );
+            }
+        }
+
+        fn read_block_repair_callback(completion: *Storage.Read) void {
+            const read = @fieldParentPtr(ReadRepair, "completion", completion);
+
+            if (read_block_valid(read.block, .{
+                .address = read.address,
+                .checksum = read.checksum,
+                .block_type = null,
+            })) {
+                // Even though this block is the block we expected to read, we can't safely
+                // cache this result:
+                // 1. Replica A write block X₁ to address X.
+                // 2. Replica A write block X₂ to address X (write is lost/misdirected).
+                // 3. Replica B requests block X₁ from replica A.
+                // It is safe for A to send back X₁, but A must not allow it to poison its cache.
```

Comment:
Maybe move that to doc comment for `read_block_repair`, and also mention that caching would be bad from perf perspective because: 

* it will fill cache with non-temporally-local blocks
* it will force an extra memcpy.